particle-iot/particle-cli
# updates the bootloader for 0.7.0 firmware on photon/p1 [ch15177]

## Problem

When running `particle update` for Device OS firmware 0.7.0 on a Photon or P1 that does not already have Wi-Fi credentials or possibly doesn't have an internet connection available, the bootloader does not get updated.  This normally happens when it connects to the Cloud in Safe Mode and the dependency requirement is fulfilled by the Safe Mode Healer.

## Solution

Change the `particle update` process for Photon / P1 only to use a special 0.7.0 system firmware that lowers bootloader dependency to v7, backs up the current user app, and applies a special "ascender" app that updates the bootloader and restores the backed up user app.

The process looks like this:
Enter DFU Mode
Run `particle udpate` as usual, but in the background:

- Save current user firmware > user_firmware_backup.bin
- Load user_firmware_backup.bin into OTA region
- Load special 0.7.0 system firmware parts
- Load 0.7.0 bootloader updater app (ascender), leave DFU mode
- Reset and run ascender app in manual mode.
- Copy bootloader directly.
- Flag OTA region to be copied to User region on next boot
- Reset and run.
- Bootloader will see and copy OTA image to User space (User firmware restored)
- Bootloader jumps to User space and runs.

## There is one known corner cases which still exists:

In the ascender app, we use a retained variable to keep track of when we’ve updated the bootloader.  That way if the app loops for some reason the ascender will not constantly be trying to write the bootloader.  A power removal during that would be fatal.  The ascender app could loop if the user runs particle update with no user app, and then runs update again with the ascender app loaded.  The ascender app would constantly be applying itself in a loop.

**Corner Case:** If you upgrade with the new method, then downgrading the bootloader and system firmware to 0.6.3, and don’t remove power before trying to upgrade again it will still have the retained memory that says it updated the bootloader already, and it won’t update it.  Also to note, if it doesn’t update the bootloader, and it gets online… since it has the new system firmware that says the bootloader >= v7 is ok, the Safe Mode Healer won’t update it either.

- So `particle update` should be run on a device that was just powered up, or if downgrading and upgrading again, be sure to remove power for 30 seconds first.

## Steps to test

- Install the prerelease `npm install -g https://github.com/particle-iot/particle-cli/releases/download/v1.30.0-rc.1/particle-cli-1.30.0-rc.1.tgz`
- Downgrade your Photon or P1 to 0.6.3 and place tinker on it (Eg. Photon)
```
particle flash --usb system-part1-0.6.3-photon.bin
particle flash --usb system-part2-0.6.3-photon.bin
particle flash --usb system-part3-0.6.3-photon.bin
particle flash --usb tinker
```
- Put device in Listening mode
- Run `particle serial inspect`
- Device should have the following signature (Eg. Photon):
```
Platform: 6 - Photon
Modules
  Bootloader module #0 - version 11, main location, 16384 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 109, main location, 262144 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #2 - version 109, main location, 262144 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 109
  User module #1 - version 2, main location, 131072 bytes max size
    UUID: 2BA4E71E840F596B812003882AAE7CA6496F1590CA4A049310AF76EAF11C943A
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #2 - version 1
  empty - factory location, 131072 bytes max size
```

- Remove power for 30 seconds if not the first time updating!
- Put device in DFU mode
- Run the CLI from the repo directory `particle-cli $ PARTICLE_CLI_DEVELOPMENT=1 node bin/particle.js update`
- Put device in Listening mode
- Run `particle serial inspect`
- Device should have the following signature (Eg. Photon):
```
Platform: 6 - Photon
Modules
  Bootloader module #0 - version 101, main location, 16384 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #1 - version 207, main location, 262144 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
  System module #2 - version 207, main location, 262144 bytes max size
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #1 - version 207
      Bootloader module #0 - version 7
  User module #1 - version 2, main location, 131072 bytes max size
    UUID: 2BA4E71E840F596B812003882AAE7CA6496F1590CA4A049310AF76EAF11C943A
    Integrity: PASS
    Address Range: PASS
    Platform: PASS
    Dependencies: PASS
      System module #2 - version 1
  empty - factory location, 131072 bytes max size
```
